mkosi/distributions/gentoo.py

```diff
@@ -211,45 +117,42 @@ def fetch_fix_stage3(self) -> None:
 
+        stage3_tar_path = state.cache_dir / stage3_tar
+
         stage3_url_path = urllib.parse.urljoin(
-            self.portage_cfg["GENTOO_MIRRORS"],
-            f"releases/{self.arch}/autobuilds/{stage3_tar}",
+            state.config.mirror, f"releases/{arch}/autobuilds/{stage3_tar}",
         )
-        stage3_tar_path = self.state.cache_dir / stage3_tar
-        stage3_tmp_extract = stage3_tar_path.with_name(stage3_tar.name + ".tmp")
-        if not stage3_tar_path.is_file():
-            log_step(f"Fetching {stage3_url_path}")
-            stage3_tar_path.parent.mkdir(parents=True, exist_ok=True)
-            urllib.request.urlretrieve(stage3_url_path, stage3_tar_path)
-
-        stage3_tmp_extract.mkdir(parents=True, exist_ok=True)
-
-        with flock(stage3_tmp_extract):
-            if not stage3_tmp_extract.joinpath(".cache_isclean").exists():
-                log_step(f"Extracting {stage3_tar.name} to {stage3_tmp_extract}")
 
+        cls.stage3_cache = state.cache_dir.joinpath("stage3")
+        if not stage3_tar_path.is_file():
+            if cls.stage3_cache.exists():
+                log_step("New stage3 is available, removing the old one")
+                unlink_try_hard(cls.stage3_cache)
+            with complete_step(f"Fetching {stage3_url_path}"):
+                stage3_tar_path.parent.mkdir(parents=True, exist_ok=True)
+                urllib.request.urlretrieve(stage3_url_path, stage3_tar_path)
+        cls.stage3_cache.mkdir(parents=True, exist_ok=True)
+        if not cls.stage3_cache.joinpath(".cache_isclean").exists():
+            with complete_step(f"Extracting {stage3_tar.name} to {cls.stage3_cache}"):
                 run([
                     "tar",
                     "--numeric-owner",
-                    "-C", stage3_tmp_extract,
+                    "-C", cls.stage3_cache,
                     "--extract",
                     "--file", stage3_tar_path,
                     "--exclude", "./dev",
+                    "--exclude", "./proc",
                 ])
 
-                unlink_try_hard(stage3_tmp_extract.joinpath("dev"))
-                unlink_try_hard(stage3_tmp_extract.joinpath("proc"))
-                unlink_try_hard(stage3_tmp_extract.joinpath("sys"))
-
-                stage3_tmp_extract.joinpath(".cache_isclean").touch()
+            cls.stage3_cache.joinpath(".cache_isclean").touch()
 
-        log_step(f"Copying {stage3_tmp_extract} to {self.root}")
-        copy_path(stage3_tmp_extract, self.root)
+        for d in ("binpkgs", "distfiles", "repos"):
+            state.cache_dir.joinpath(d).mkdir(exist_ok=True)
 
-    def set_useflags(self) -> None:
-        package_use = self.portage_cfg_dir / "package.use"
-        package_use.mkdir(exist_ok=True)
+        config = state.pkgmngr / "etc/portage"
 
+        package_use = config / "package.use"
+        package_use.mkdir(parents=True, exist_ok=True)
         package_use.joinpath("systemd").write_text(
```

Comment:
This and some other instances further below clobber the files in `state.pkgmngr`. The idea is for users to be able to pass in their own files. Is there a way to make this work for Gentoo? In the other distros we try hard to not clobber things and instead pass arguments on the cmdline to overwrite config files.

Reply:
> This and some other instances further below clobber the files in state.pkgmngr.

that's the effect of gentoo being bare minimum whereas other distros, it seems to me, come with a canonical base config which is probably the side effect of being debian/fedora offshoots basically and following the _upstream_ to the letter.
 
> The idea is for users to be able to pass in their own files.

i thought and used, `skeleton/extra` for that in the past.

> Is there a way to make this work for Gentoo? In the other distros we try hard to not clobber things and instead pass arguments on the cmdline to overwrite config files.

i see. please see my question to @DaanDeMeyer above, regarding `install_package_manager_trees()`. that happens before things are handed over to the `installer()` instance. i revisit these as in  https://github.com/systemd/mkosi/issues/1640

Reply:
> i see. please see my question to @DaanDeMeyer above, regarding `install_package_manager_trees()`. that happens before things are handed over to the `installer()` instance. i revisit these as in #1640

https://github.com/systemd/mkosi/pull/1570#discussion_r1238413322

Reply:
> This and some other instances further below clobber the files in `state.pkgmngr`.

see this commit: `gentoo: don't clobber pkgmngr, rework repositories`